packages/sdk/client-services/src/packlets/vault/localstorage-resource-manager.ts

```diff
@@ -0,0 +1,78 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { v4 as uuid } from 'uuid';
+
+import { log, logInfo } from '@dxos/log';
+import { MaybePromise } from '@dxos/util';
+
+export type LocalStorageResourceManagerParams = {
+  key: string;
+  id?: string;
+  onAcquired?: () => MaybePromise<void>;
+  onReleased?: () => MaybePromise<void>;
+};
+
+export class LocalStorageResourceManager {
+  @logInfo
+  readonly key: string;
+
+  @logInfo
+  readonly id: string;
+
+  private readonly _onAcquired?: () => MaybePromise<void>;
+  private readonly _onReleased?: () => MaybePromise<void>;
+
+  constructor({ key, id, onAcquired, onReleased }: LocalStorageResourceManagerParams) {
+    this.key = key;
+    this.id = id ?? uuid();
+    this._onAcquired = onAcquired;
+    this._onReleased = onReleased;
+
+    window.addEventListener('storage', this._onStorageEvent.bind(this));
+  }
+
+  hasAcquired() {
+    const id = this._currentId();
+    return id === this.id;
+  }
+
+  async acquire() {
+    log('acquiring...', { key: this.key, id: this.id });
+    localStorage.setItem(this.key, JSON.stringify({ id: this.id, time: Date.now() }));
```

Comment:
this should check whether some other tab is holding the lock

Reply:
This is setup like this along the lines of what we discussed, where any tab would be able to resume usage (aka acquire access to the vault). That being said, I think you're right that it needs to be more aware of the tabs access. Maybe something like the `acquire` method doesn't resolve until an existing tab has finished releasing? So they would need to communicate via a second key likely.